feat(perf-issues) Update perf group creation

udameli opened the pull request:
This PR enables performance group creation/update. For fingerprints coming out of job["performance_problems"] an existing group gets updated or a new group created. Group type is set based on the output of the performance detection. Performance issue creation is rate limited to ensure that we don't create too many groups at once.

Commit:
initial performance issue creation

WIP fixes

reset defaults

remove testing logic

consolidate duplicate logic

consolidated more group creation logic

consolidate group creation logic

return perf problems from detection

fix imports

address comments

start updating the perf grouping logic

add a new group type

wip

update group creation

use groups instead of group on jobs

use `GroupInfo` in both errors and transations save event pipeline

don't need `raw`

return correct fingerprint

add correct option

update tests to use new option

fix test
unbreak transaction saving
make sure we use correct Sequence

add per project option

wedamija commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -196,9 +196,11 @@ def prepare_problem_for_grouping(
 def fingerprint_group(transaction_name, span_op, hash, problem_class):
-    signature = (str(transaction_name) + str(span_op) + str(hash)).encode("utf-8")
+    signature = (str(transaction_name) + str(span_op) + str(hash) + str(problem_class)).encode(
+        "utf-8"
+    )
     full_fingerprint = hashlib.sha1(signature).hexdigest()
-    return f"1-{problem_class}-{full_fingerprint}"
+    return full_fingerprint
```
@udameli Not sure if i messed up the rebase here, is this code correct?

udameli replied:
yeah it should be `f"1-{problem_class}-{full_fingerprint}"`. I pushed the fix 👍

udameli replied:
re-pushed. should be fixed now

udameli commented on src/sentry/event_manager.py:
```diff
@@ -1928,6 +1938,106 @@ def _detect_performance_problems(jobs, projects):
+        rate = options.get("performance.issues.all.problem-creation")
+
+        # More granular, per-project option
+        per_project_rate = projects[idx].get_option("sentry:performance_issue_creation_rate")
```
this project option was added here https://github.com/getsentry/sentry/pull/38470

udameli commented on src/sentry/event_manager.py:
```diff
@@ -1928,6 +1938,106 @@ def _detect_performance_problems(jobs, projects):
+        rate = options.get("performance.issues.all.problem-creation")
+
+        # More granular, per-project option
+        per_project_rate = projects[idx].get_option("sentry:performance_issue_creation_rate")
```
I'm making an assumption here that jobs and projects with the same index are related. does that make sense here? @wedamija

wedamija replied:
Projects is a dict https://github.com/getsentry/sentry/blob/c279356cca47149a1d861bfc1abba910b6df0b07/src/sentry/event_manager.py#L361
You'd have to get the project id associated with the job instead.

udameli replied:
ah i saw below that we're already getting a `project` below so i moved those lines higher up
```
 event = job["event"]
 project = event.project
```

wedamija committed:
add basics of ratelimiting

wedamija commented on src/sentry/event_manager.py:
```diff
@@ -109,6 +109,11 @@
+issue_rate_limiter = RedisSlidingWindowRateLimiter(
+    **settings.SENTRY_PERFORMANCE_ISSUES_RATE_LIMITER_OPTIONS
+)
+PERFORMANCE_ISSUE_QUOTA = Quota(3600, 60, 60)
```
This means 60 issues per hour, with a sliding window granularity of a minute. just dummy values, not sure what we'll really use

Commit:
use project on the event

Commit:
set project options

wedamija committed:
fixes for `_save_aggregate_performance`
 - problem fingerprints need to be md5ed
 - new_grouphashes converted to a list before limited by rate limited
 - GroupHash creation needs to pass kwargs
 - Need to pass all groups to `event.groups` instead of `group_ids` since we removed `group_ids`

Added hack try/except around report_metrics_for_detectors since it's failing with the span
currently. Should investigate

wedamija committed:
Update tests to check groups are created with correct type
simplify rate limit checking with defaults

Commit:
set group type on creation

wedamija committed:
make sure metadata, title etc are properly set

Commit:
fix tests

udameli commented on src/sentry/event_manager.py:
```diff
@@ -1935,6 +1945,122 @@ def _detect_performance_problems(jobs, projects):
+                                if problem.fingerprint == new_grouphash
+                            ),
+                            None,
+                        )
```
this looks up group type based on a group hash

Commit:
update title

Commit:
update title

udameli commented on src/sentry/event_manager.py:
```diff
@@ -1935,6 +1945,133 @@ def _detect_performance_problems(jobs, projects):
+                        metric_tags["create_group_transaction.outcome"] = "no_group"
+
+                        problem = next(
+                            (
```
gotta restructure `performance_problems` so looking up problems by fingerprint is faster

Commit:
fix

Commit:
fix for sure this time

wedamija committed:
lookup perf problems by hash efficiently

wedamija committed:
add test for existing groups

wedamija committed:
make sure we only store unprocessed for error events

wedamija committed:
oops options.get doesn't support a default

Commit:
track number of dropped new groups

gggritso commented on src/sentry/event_manager.py:
```diff
@@ -1935,6 +1944,123 @@ def _detect_performance_problems(jobs, projects):
+
+                        problem = performance_problems_by_fingerprint[new_grouphash]
+                        kwargs["type"] = problem.type.value
+                        kwargs["data"]["metadata"]["title"] = f"N+1 Query:{problem.desc}"
```
Oh! Maybe the problem detectors should return a string for the problem type?

udameli replied:
yeah that'd work here better 👍 gonna hard-code it for now since we're only returning one type of perf problems

wedamija replied:
I think we should just have a mapping of `GroupType` -> `str` somewhere, since the `PerformanceProblem` already has that reference. Performance detection doesn't really need to know how we display a given performance problem

gggritso commented on src/sentry/event_manager.py:
```diff
@@ -1935,6 +1944,123 @@ def _detect_performance_problems(jobs, projects):
+
+            performance_problems = job["performance_problems"]
+            for problem in performance_problems:
+                problem.fingerprint = md5(problem.fingerprint.encode("utf-8")).hexdigest()
```
It it necessary to `hexdigest` this? It might be easier for us to debug fingerprinting in the short term if we can see the original fingerprint string, as returned from the detector

udameli replied:
yeah I've also just realized today that fingerprint has to be a 32 chars long

wedamija replied:
Our fingerprints are typically just hashes and we don't expect to be able to extra information from them. We could potentially store the fingerprints in the `Group`'s data field for debugging.

gggritso reviewed (approved):
B-e-a-utiful! Just two small Qs

Commit:
mock sdk_span.containing_transaction

Commit:
remove nesting of withs

udameli commented on tests/sentry/performance_issues/test_performance_issues.py:
```diff
@@ -36,71 +37,76 @@ class EventManagerTest(TestCase, EventManagerTestMixin):
-            assert group.issue_category == GroupCategory.PERFORMANCE
-            assert group.issue_type == GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES
+        with mock.patch("sentry_sdk.tracing.Span.containing_transaction"):
+            with self.feature("projects:performance-suspect-spans-ingestion"):
```
ty, that's so much better. fixed 👍

Commit:
Merge branch 'master' into du/feat/update-perf-group-creation
